compiler/frontend/src/org/jetbrains/kotlin/effectsystem/adapters/CallTreeBuilder.kt

```diff
@@ -0,0 +1,194 @@
+        val value: Any? = compileTimeConstant.getValue(type)
+        return CTConstant(ValueIdsFactory.idForConstant(value), type, value)
+    }
+
+    override fun visitSimpleNameExpression(expression: KtSimpleNameExpression, data: Unit): CTNode {
+        // TODO: make proper resolving
+        if (expression.text == "Unit") return CTConstant(UNIT_ID, DefaultBuiltIns.Instance.unitType, Unit)
+        return tryCreateVariable(expression)
+    }
+
+    override fun visitParenthesizedExpression(expression: KtParenthesizedExpression, data: Unit): CTNode {
+        val deparenthesized = KtPsiUtil.safeDeparenthesize(expression)
+        return if (deparenthesized == expression) UNKNOWN_CALL else deparenthesized.accept(this, data)
+    }
+
+    override fun visitUnaryExpression(expression: KtUnaryExpression, data: Unit): CTCall {
+        tryGetCachedCall(expression)?.let { return it }
+
+        val argNode = expression.baseExpression?.accept(this, data) ?: return UNKNOWN_CALL
+        return when (expression.operationToken) {
+            KtTokens.EXCL -> CTCall(NotFunctor(), listOf(argNode))
+            else -> return UNKNOWN_CALL
+        }
+    }
+
+    override fun visitDotQualifiedExpression(expression: KtDotQualifiedExpression, data: Unit): CTCall {
+        tryGetCachedCall(expression)?.let { return it }
+
+        val receiver = expression.receiverExpression.accept(this, data)
+
+        val resolvedCall = expression.selectorExpression.getResolvedCall(bindingContext) ?: return UNKNOWN_CALL
+        val argNodes = resolvedCall.valueArgumentsByIndex?.map {
+            (it as? ExpressionValueArgument)?.valueArgument?.getArgumentExpression()?.accept(this, data) ?: return UNKNOWN_CALL
+        } ?: return UNKNOWN_CALL
+
+        val functor = functorResolver.resolveFunctor(resolvedCall) ?: return UNKNOWN_CALL
+        return CTCall(functor, listOf(receiver) + argNodes)
+    }
+
+    override fun visitThisExpression(expression: KtThisExpression, data: Unit?): CTNode {
+        val dfv = expression.createDataFlowValue() ?: return UNKNOWN_CALL // Could be possible for unavailable/incorrect this
+        return CTVariable(ValueIdsFactory.dfvBased(dfv), dfv.type)
+    }
+
+    override fun visitClassLiteralExpression(expression: KtClassLiteralExpression, data: Unit?): CTNode = tryCreateVariable(expression)
+
+    override fun visitLabeledExpression(expression: KtLabeledExpression, data: Unit): CTNode = expression.baseExpression?.accept(this, data) ?: UNKNOWN_CALL
+
+    override fun visitBinaryExpression(expression: KtBinaryExpression, data: Unit): CTCall {
+        tryGetCachedCall(expression)?.let { return it }
+
+        val leftNode = expression.left?.accept(this, data) ?: return UNKNOWN_CALL
+        val rightNode = expression.right?.accept(this, data) ?: return UNKNOWN_CALL
+
+        val functor = when (expression.operationToken) {
+            KtTokens.EQEQ, KtTokens.EQEQEQ -> {
```

Comment:
I think, everything is working correctly now, but only due to the very limited amount of cases where equality actually supported (only `Boolean` to `true`/`false` and comparisons with `null`). Still, it can potentially backfire if we will decide to cover more cases, so I think you're right, we should treat identity comparison more carefully (with separate functor, probably)